server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java

```diff
@@ -98,7 +98,26 @@ interface StoredFields {
 
     RowStrideReader rowStrideReader(LeafReaderContext context) throws IOException;
 
-    StoredFieldsSpec rowStrideStoredFieldSpec();
+    record FieldsSpec(StoredFieldsSpec storedFieldsSpec, IgnoredFieldsSpec ignoredFieldsSpec) {
```

Comment:
Would it make sense to extend `StoredFieldsSpec` to know about ignored fields, instead of having this new abstraction on top?  I think that would make a lot of the logic here easier to follow.

Reply:
I was reluctant to do so because StoredFieldsSpec is used in a few other places, and I only need the IgnoredFieldsSpec in the ESQL blockloader context.

Reply:
I'm also leaning towards keeping `IgnoredFieldsSpec` in this PR.

Reply:
After taking a second look at this, if we're able to add `IgnoredFieldsSpec` to `StoredFieldsSpec` and make it optional then I think it would reduce the size of this PR? By adding a constructor to `StoredFieldSpec`:

```
public StoredFieldsSpec(boolean requiresSource, boolean requiresMetadata, Set<String> requiredStoredFields) {
        this(requiresSource, requiresMetadata, requiredStoredFields, IgnoredFieldsSpec.NONE);
    }
```

(the `StoredFieldsSpec#merge(...)` method also needs to be updated,  but from there we can delegate to IgnoredFieldsSpec)

`StoredFieldSpec` is used in fetch phase and value fetchers, but I think we don't need to make any changes in these areas of the code base.